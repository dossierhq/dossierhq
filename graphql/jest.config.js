module.exports = {
  displayName: 'graphql',
  preset: 'ts-jest',
  testEnvironment: 'node',
  globalSetup: './test/setup.ts',
  globalTeardown: './test/teardown.ts',
};
