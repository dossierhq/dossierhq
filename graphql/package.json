{
  "name": "@jonasb/datadata-graphql",
  "version": "0.1.1-7",
  "description": "",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "exports": {
    ".": "./lib/index.js"
  },
  "scripts": {
    "build": "heft build",
    "build:watch": "heft build --watch",
    "ci:check": "npm run test && npm run prettier:check",
    "clean": "heft clean",
    "db:cli": "./scripts/db-cli.sh",
    "db:migrate": "./scripts/db-migrate.sh",
    "eslint:check": "eslint .",
    "eslint:fix": "eslint --fix .",
    "prettier:check": "prettier --check .",
    "prettier:fix": "prettier --write .",
    "test": "heft test",
    "test:update-snapshot": "heft test --update-snapshots",
    "test:watch": "heft test --watch",
    "ts:check": "tsc --noEmit",
    "ts:check:watch": "tsc --noEmit --watch",
    "upgrade-dependencies": "npx npm-check-updates -u --packageFile ./package.json",
    "upgrade-dependencies:check": "npx npm-check-updates"
  },
  "author": "",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/jonasb/datadata.git"
  },
  "devDependencies": {
    "@jonasb/datadata-postgres": "0.1.1-7",
    "@jonasb/datadata-server": "0.1.1-7",
    "@jonasb/datadata-testing-utils": "0.1.1-7",
    "@rushstack/heft": "~0.34.8",
    "@rushstack/heft-jest-plugin": "~0.1.14",
    "@types/heft-jest": "1.0.2",
    "@types/uuid": "^8.3.0",
    "@typescript-eslint/eslint-plugin": "^4.24.0",
    "@typescript-eslint/parser": "^4.24.0",
    "eslint": "^7.26.0",
    "prettier": "2.3.0",
    "typescript": "^4.3.5",
    "uuid": "^8.3.2"
  },
  "dependencies": {
    "@jonasb/datadata-core": "0.1.1-7",
    "graphql": "~15.5.0"
  }
}
