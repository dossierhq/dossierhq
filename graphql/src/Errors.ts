export class NotAuthenticatedError extends Error {
  name = 'NotAuthenticatedError';
  constructor(message = 'Not authenticated') {
    super(message);
  }
}
