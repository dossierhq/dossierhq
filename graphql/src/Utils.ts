import type { AdminClient } from '@datadata/core';
import type { SessionContext } from '@datadata/server';
import type { SessionGraphQLContext } from '.';

export function getSessionContext<TContext extends SessionGraphQLContext>(
  context: TContext
): SessionContext {
  if (context.context.isError()) {
    throw context.context.toError();
  }
  return context.context.value;
}

export function getAdminClient<TContext extends SessionGraphQLContext>(
  context: TContext
): AdminClient {
  if (context.adminClient.isError()) {
    throw context.adminClient.toError();
  }
  return context.adminClient.value;
}
