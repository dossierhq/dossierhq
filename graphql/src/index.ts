export type { SessionGraphQLContext } from './GraphQLSchemaGenerator';

export { GraphQLSchemaGenerator } from './GraphQLSchemaGenerator';
