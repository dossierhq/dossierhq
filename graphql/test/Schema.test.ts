import { EntityFieldType, Schema } from '@datadata/core';
import type { SchemaSpecification } from '@datadata/core';
import { graphql, printSchema } from 'graphql';
import { GraphQLSchemaGenerator } from '../src/GraphQLSchemaGenerator';

function describeGeneratedSchema(schemaSpec: SchemaSpecification) {
  const generator = new GraphQLSchemaGenerator(new Schema(schemaSpec));
  const graphQLSchema = generator.buildSchema();
  return printSchema(graphQLSchema);
}

async function querySchema(schemaSpec: SchemaSpecification, query: string) {
  const generator = new GraphQLSchemaGenerator(new Schema(schemaSpec));
  const graphQLSchema = generator.buildSchema();
  return await graphql(graphQLSchema, query);
}

describe('Empty schema spec', () => {
  const schemaSpec = { entityTypes: {} };
  test('Generated QL schema', () => {
    const result = describeGeneratedSchema(schemaSpec);
    expect(result).toMatchSnapshot();
  });
});

describe('One empty entity type schema spec', () => {
  const schemaSpec = { entityTypes: { Foo: { fields: [] } } };
  test('Generated QL schema', () => {
    const result = describeGeneratedSchema(schemaSpec);
    expect(result).toMatchSnapshot();
  });

  test('Ensure Node matches Global Object Identification spec', async () => {
    // From https://graphql.org/learn/global-object-identification/
    const query = `{
      __type(name: "Node") {
        name
        kind
        fields {
          name
          type {
            kind
            ofType {
              name
              kind
            }
          }
        }
      }
    }`;
    const expected = {
      data: {
        __type: {
          name: 'Node',
          kind: 'INTERFACE',
          fields: [
            {
              name: 'id',
              type: {
                kind: 'NON_NULL',
                ofType: {
                  name: 'ID',
                  kind: 'SCALAR',
                },
              },
            },
          ],
        },
      },
    };
    const result = await querySchema(schemaSpec, query);
    expect(result).toEqual(expected);
  });

  test('Ensure node() matches Global Object Identification spec', async () => {
    // From https://graphql.org/learn/global-object-identification/
    const query = `{
      __schema {
        queryType {
          fields {
            name
            type {
              name
              kind
            }
            args {
              name
              type {
                kind
                ofType {
                  name
                  kind
                }
              }
            }
          }
        }
      }
    }`;
    const expected = {
      data: {
        __schema: {
          queryType: {
            fields: [
              // This array may have other entries
              {
                name: 'node',
                type: {
                  name: 'Node',
                  kind: 'INTERFACE',
                },
                args: [
                  {
                    name: 'id',
                    type: {
                      kind: 'NON_NULL',
                      ofType: {
                        name: 'ID',
                        kind: 'SCALAR',
                      },
                    },
                  },
                ],
              },
            ],
          },
        },
      },
    };
    const result = await querySchema(schemaSpec, query);

    // remove all fields except 'node'
    const queryType = result.data?.__schema.queryType as { fields: { name: string }[] };
    queryType.fields = queryType.fields.filter((x) => x.name === 'node');

    expect(result).toEqual(expected);
  });
});

describe('Two entity types with reference schema spec', () => {
  const schemaSpec = {
    entityTypes: {
      Foo: { fields: [{ name: 'fooField', type: EntityFieldType.String }] },
      Bar: {
        fields: [
          { name: 'barField1', type: EntityFieldType.String },
          { name: 'barField2', type: EntityFieldType.Reference },
        ],
      },
    },
  };
  test('Generated QL schema', () => {
    const result = describeGeneratedSchema(schemaSpec);
    expect(result).toMatchSnapshot();
  });
});
