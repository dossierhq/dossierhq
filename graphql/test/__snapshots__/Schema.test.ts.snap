// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
}
"
`;

exports[`List of strings, locations and references schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

enum EntityType {
  Foo
  Bar
}

interface Entity implements Node {
  id: ID!
  _name: String!
}

type RichText {
  blocksJson: String!
  entities: [Entity]
}

type Location {
  lat: Float!
  lng: Float!
}

input LocationInput {
  lat: Float!
  lng: Float!
}

input BoundingBoxInput {
  minLat: Float!
  maxLat: Float!
  minLng: Float!
  maxLng: Float!
}

type Foo implements Node & Entity {
  id: ID!
  _name: String!
  strings: [String!]
  locations: [Location!]
  bars: [Bar!]
}

type Bar implements Node & Entity {
  id: ID!
  _name: String!
}

enum EntityPublishState {
  draft
  published
  modified
  withdrawn
  archived
}

interface AdminEntity {
  id: ID!
  _name: String!
  _type: EntityType!
  _version: Int!
  _publishState: EntityPublishState!
}

type AdminEntityEdge {
  node: AdminEntity
  cursor: String!
}

type AdminEntityConnection {
  pageInfo: PageInfo!
  edges: [AdminEntityEdge]
  totalCount: Int!
}

input AdminQueryInput {
  entityTypes: [EntityType]
  referencing: ID
  boundingBox: BoundingBoxInput
  order: String
  text: String
}

input AdminReferenceInput {
  id: ID!
}

input EntityVersionInput {
  id: ID!
  version: Int!
}

type AdminRichText {
  blocksJson: String!
  entities: [AdminEntity]
}

input AdminRichTextInput {
  blocksJson: String!
}

type EntityVersionInfo {
  version: Int!
  published: Boolean!
  createdBy: ID!
  createdAt: String!
}

type EntityHistory {
  id: ID!
  versions: [EntityVersionInfo]!
}

type PublishEvent {
  version: Int
  publishedBy: ID!
  publishedAt: String!
}

type PublishHistory {
  id: ID!
  events: [PublishEvent]!
}

type EntityPublishPayload {
  id: ID!
}

type AdminFoo implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
  strings: [String!]
  locations: [Location!]
  bars: [AdminBar!]
}

input AdminFooCreateInput {
  id: ID
  _type: EntityType
  _name: String!
  strings: [String!]
  locations: [LocationInput!]
  bars: [AdminReferenceInput!]
}

input AdminFooUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
  strings: [String!]
  locations: [LocationInput!]
  bars: [AdminReferenceInput!]
}

type AdminBar implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
}

input AdminBarCreateInput {
  id: ID
  _type: EntityType
  _name: String!
}

input AdminBarUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
  adminEntity(id: ID!, version: Int): AdminEntity
  adminEntities(ids: [ID!]!): [AdminEntity]
  adminSearchEntities(query: AdminQueryInput, first: Int, after: String, last: Int, before: String): AdminEntityConnection
  entityHistory(id: ID!): EntityHistory
  publishHistory(id: ID!): PublishHistory
}

type Mutation {
  publishEntities(entities: [EntityVersionInput!]!): [EntityPublishPayload!]
  unpublishEntities(ids: [ID!]!): [EntityPublishPayload!]
  archiveEntity(id: ID!): EntityPublishPayload
  unarchiveEntity(id: ID!): EntityPublishPayload
  createFooEntity(entity: AdminFooCreateInput!): AdminFoo
  updateFooEntity(entity: AdminFooUpdateInput!): AdminFoo
  createBarEntity(entity: AdminBarCreateInput!): AdminBar
  updateBarEntity(entity: AdminBarUpdateInput!): AdminBar
}
"
`;

exports[`Multiple references with entityTypes schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

enum EntityType {
  Foo
  Bar
  Baz
}

interface Entity implements Node {
  id: ID!
  _name: String!
}

type RichText {
  blocksJson: String!
  entities: [Entity]
}

type Location {
  lat: Float!
  lng: Float!
}

input LocationInput {
  lat: Float!
  lng: Float!
}

input BoundingBoxInput {
  minLat: Float!
  maxLat: Float!
  minLng: Float!
  maxLng: Float!
}

type Foo implements Node & Entity {
  id: ID!
  _name: String!
  noMeansAll: Entity
  bar: Bar
  bazBar: _BarOrBaz
  fooBarBaz: _BarOrBazOrFoo
  barBarBar: Bar
}

type Bar implements Node & Entity {
  id: ID!
  _name: String!
}

type Baz implements Node & Entity {
  id: ID!
  _name: String!
}

enum EntityPublishState {
  draft
  published
  modified
  withdrawn
  archived
}

interface AdminEntity {
  id: ID!
  _name: String!
  _type: EntityType!
  _version: Int!
  _publishState: EntityPublishState!
}

type AdminEntityEdge {
  node: AdminEntity
  cursor: String!
}

type AdminEntityConnection {
  pageInfo: PageInfo!
  edges: [AdminEntityEdge]
  totalCount: Int!
}

input AdminQueryInput {
  entityTypes: [EntityType]
  referencing: ID
  boundingBox: BoundingBoxInput
  order: String
  text: String
}

input AdminReferenceInput {
  id: ID!
}

input EntityVersionInput {
  id: ID!
  version: Int!
}

type AdminRichText {
  blocksJson: String!
  entities: [AdminEntity]
}

input AdminRichTextInput {
  blocksJson: String!
}

type EntityVersionInfo {
  version: Int!
  published: Boolean!
  createdBy: ID!
  createdAt: String!
}

type EntityHistory {
  id: ID!
  versions: [EntityVersionInfo]!
}

type PublishEvent {
  version: Int
  publishedBy: ID!
  publishedAt: String!
}

type PublishHistory {
  id: ID!
  events: [PublishEvent]!
}

type EntityPublishPayload {
  id: ID!
}

type AdminFoo implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
  noMeansAll: AdminEntity
  bar: AdminBar
  bazBar: _AdminBarOrBaz
  fooBarBaz: _AdminBarOrBazOrFoo
  barBarBar: AdminBar
}

input AdminFooCreateInput {
  id: ID
  _type: EntityType
  _name: String!
  noMeansAll: AdminReferenceInput
  bar: AdminReferenceInput
  bazBar: AdminReferenceInput
  fooBarBaz: AdminReferenceInput
  barBarBar: AdminReferenceInput
}

input AdminFooUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
  noMeansAll: AdminReferenceInput
  bar: AdminReferenceInput
  bazBar: AdminReferenceInput
  fooBarBaz: AdminReferenceInput
  barBarBar: AdminReferenceInput
}

type AdminBar implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
}

input AdminBarCreateInput {
  id: ID
  _type: EntityType
  _name: String!
}

input AdminBarUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
}

type AdminBaz implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
}

input AdminBazCreateInput {
  id: ID
  _type: EntityType
  _name: String!
}

input AdminBazUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
}

enum _BarOrBaz {
  Bar
  Baz
}

enum _BarOrBazOrFoo {
  Bar
  Baz
  Foo
}

enum _AdminBarOrBaz {
  AdminBar
  AdminBaz
}

enum _AdminBarOrBazOrFoo {
  AdminBar
  AdminBaz
  AdminFoo
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
  adminEntity(id: ID!, version: Int): AdminEntity
  adminEntities(ids: [ID!]!): [AdminEntity]
  adminSearchEntities(query: AdminQueryInput, first: Int, after: String, last: Int, before: String): AdminEntityConnection
  entityHistory(id: ID!): EntityHistory
  publishHistory(id: ID!): PublishHistory
}

type Mutation {
  publishEntities(entities: [EntityVersionInput!]!): [EntityPublishPayload!]
  unpublishEntities(ids: [ID!]!): [EntityPublishPayload!]
  archiveEntity(id: ID!): EntityPublishPayload
  unarchiveEntity(id: ID!): EntityPublishPayload
  createFooEntity(entity: AdminFooCreateInput!): AdminFoo
  updateFooEntity(entity: AdminFooUpdateInput!): AdminFoo
  createBarEntity(entity: AdminBarCreateInput!): AdminBar
  updateBarEntity(entity: AdminBarUpdateInput!): AdminBar
  createBazEntity(entity: AdminBazCreateInput!): AdminBaz
  updateBazEntity(entity: AdminBazUpdateInput!): AdminBaz
}
"
`;

exports[`One empty entity type schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

enum EntityType {
  Foo
}

interface Entity implements Node {
  id: ID!
  _name: String!
}

type RichText {
  blocksJson: String!
  entities: [Entity]
}

type Location {
  lat: Float!
  lng: Float!
}

input LocationInput {
  lat: Float!
  lng: Float!
}

input BoundingBoxInput {
  minLat: Float!
  maxLat: Float!
  minLng: Float!
  maxLng: Float!
}

type Foo implements Node & Entity {
  id: ID!
  _name: String!
}

enum EntityPublishState {
  draft
  published
  modified
  withdrawn
  archived
}

interface AdminEntity {
  id: ID!
  _name: String!
  _type: EntityType!
  _version: Int!
  _publishState: EntityPublishState!
}

type AdminEntityEdge {
  node: AdminEntity
  cursor: String!
}

type AdminEntityConnection {
  pageInfo: PageInfo!
  edges: [AdminEntityEdge]
  totalCount: Int!
}

input AdminQueryInput {
  entityTypes: [EntityType]
  referencing: ID
  boundingBox: BoundingBoxInput
  order: String
  text: String
}

input AdminReferenceInput {
  id: ID!
}

input EntityVersionInput {
  id: ID!
  version: Int!
}

type AdminRichText {
  blocksJson: String!
  entities: [AdminEntity]
}

input AdminRichTextInput {
  blocksJson: String!
}

type EntityVersionInfo {
  version: Int!
  published: Boolean!
  createdBy: ID!
  createdAt: String!
}

type EntityHistory {
  id: ID!
  versions: [EntityVersionInfo]!
}

type PublishEvent {
  version: Int
  publishedBy: ID!
  publishedAt: String!
}

type PublishHistory {
  id: ID!
  events: [PublishEvent]!
}

type EntityPublishPayload {
  id: ID!
}

type AdminFoo implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
}

input AdminFooCreateInput {
  id: ID
  _type: EntityType
  _name: String!
}

input AdminFooUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
  adminEntity(id: ID!, version: Int): AdminEntity
  adminEntities(ids: [ID!]!): [AdminEntity]
  adminSearchEntities(query: AdminQueryInput, first: Int, after: String, last: Int, before: String): AdminEntityConnection
  entityHistory(id: ID!): EntityHistory
  publishHistory(id: ID!): PublishHistory
}

type Mutation {
  publishEntities(entities: [EntityVersionInput!]!): [EntityPublishPayload!]
  unpublishEntities(ids: [ID!]!): [EntityPublishPayload!]
  archiveEntity(id: ID!): EntityPublishPayload
  unarchiveEntity(id: ID!): EntityPublishPayload
  createFooEntity(entity: AdminFooCreateInput!): AdminFoo
  updateFooEntity(entity: AdminFooUpdateInput!): AdminFoo
}
"
`;

exports[`Rich text schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

enum EntityType {
  Foo
}

interface Entity implements Node {
  id: ID!
  _name: String!
}

type RichText {
  blocksJson: String!
  entities: [Entity]
}

type Location {
  lat: Float!
  lng: Float!
}

input LocationInput {
  lat: Float!
  lng: Float!
}

input BoundingBoxInput {
  minLat: Float!
  maxLat: Float!
  minLng: Float!
  maxLng: Float!
}

type Foo implements Node & Entity {
  id: ID!
  _name: String!
  body: RichText
}

enum EntityPublishState {
  draft
  published
  modified
  withdrawn
  archived
}

interface AdminEntity {
  id: ID!
  _name: String!
  _type: EntityType!
  _version: Int!
  _publishState: EntityPublishState!
}

type AdminEntityEdge {
  node: AdminEntity
  cursor: String!
}

type AdminEntityConnection {
  pageInfo: PageInfo!
  edges: [AdminEntityEdge]
  totalCount: Int!
}

input AdminQueryInput {
  entityTypes: [EntityType]
  referencing: ID
  boundingBox: BoundingBoxInput
  order: String
  text: String
}

input AdminReferenceInput {
  id: ID!
}

input EntityVersionInput {
  id: ID!
  version: Int!
}

type AdminRichText {
  blocksJson: String!
  entities: [AdminEntity]
}

input AdminRichTextInput {
  blocksJson: String!
}

type EntityVersionInfo {
  version: Int!
  published: Boolean!
  createdBy: ID!
  createdAt: String!
}

type EntityHistory {
  id: ID!
  versions: [EntityVersionInfo]!
}

type PublishEvent {
  version: Int
  publishedBy: ID!
  publishedAt: String!
}

type PublishHistory {
  id: ID!
  events: [PublishEvent]!
}

type EntityPublishPayload {
  id: ID!
}

type AdminFoo implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
  body: AdminRichText
}

input AdminFooCreateInput {
  id: ID
  _type: EntityType
  _name: String!
  body: AdminRichTextInput
}

input AdminFooUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
  body: AdminRichTextInput
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
  adminEntity(id: ID!, version: Int): AdminEntity
  adminEntities(ids: [ID!]!): [AdminEntity]
  adminSearchEntities(query: AdminQueryInput, first: Int, after: String, last: Int, before: String): AdminEntityConnection
  entityHistory(id: ID!): EntityHistory
  publishHistory(id: ID!): PublishHistory
}

type Mutation {
  publishEntities(entities: [EntityVersionInput!]!): [EntityPublishPayload!]
  unpublishEntities(ids: [ID!]!): [EntityPublishPayload!]
  archiveEntity(id: ID!): EntityPublishPayload
  unarchiveEntity(id: ID!): EntityPublishPayload
  createFooEntity(entity: AdminFooCreateInput!): AdminFoo
  updateFooEntity(entity: AdminFooUpdateInput!): AdminFoo
}
"
`;

exports[`Two entity types with reference schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

enum EntityType {
  Foo
  Bar
}

interface Entity implements Node {
  id: ID!
  _name: String!
}

type RichText {
  blocksJson: String!
  entities: [Entity]
}

type Location {
  lat: Float!
  lng: Float!
}

input LocationInput {
  lat: Float!
  lng: Float!
}

input BoundingBoxInput {
  minLat: Float!
  maxLat: Float!
  minLng: Float!
  maxLng: Float!
}

type Foo implements Node & Entity {
  id: ID!
  _name: String!
  fooField: String
}

type Bar implements Node & Entity {
  id: ID!
  _name: String!
  barField1: String
  barField2: Entity
  barField3: Location
}

enum EntityPublishState {
  draft
  published
  modified
  withdrawn
  archived
}

interface AdminEntity {
  id: ID!
  _name: String!
  _type: EntityType!
  _version: Int!
  _publishState: EntityPublishState!
}

type AdminEntityEdge {
  node: AdminEntity
  cursor: String!
}

type AdminEntityConnection {
  pageInfo: PageInfo!
  edges: [AdminEntityEdge]
  totalCount: Int!
}

input AdminQueryInput {
  entityTypes: [EntityType]
  referencing: ID
  boundingBox: BoundingBoxInput
  order: String
  text: String
}

input AdminReferenceInput {
  id: ID!
}

input EntityVersionInput {
  id: ID!
  version: Int!
}

type AdminRichText {
  blocksJson: String!
  entities: [AdminEntity]
}

input AdminRichTextInput {
  blocksJson: String!
}

type EntityVersionInfo {
  version: Int!
  published: Boolean!
  createdBy: ID!
  createdAt: String!
}

type EntityHistory {
  id: ID!
  versions: [EntityVersionInfo]!
}

type PublishEvent {
  version: Int
  publishedBy: ID!
  publishedAt: String!
}

type PublishHistory {
  id: ID!
  events: [PublishEvent]!
}

type EntityPublishPayload {
  id: ID!
}

type AdminFoo implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
  fooField: String
}

input AdminFooCreateInput {
  id: ID
  _type: EntityType
  _name: String!
  fooField: String
}

input AdminFooUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
  fooField: String
}

type AdminBar implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
  barField1: String
  barField2: AdminEntity
  barField3: Location
}

input AdminBarCreateInput {
  id: ID
  _type: EntityType
  _name: String!
  barField1: String
  barField2: AdminReferenceInput
  barField3: LocationInput
}

input AdminBarUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
  barField1: String
  barField2: AdminReferenceInput
  barField3: LocationInput
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
  adminEntity(id: ID!, version: Int): AdminEntity
  adminEntities(ids: [ID!]!): [AdminEntity]
  adminSearchEntities(query: AdminQueryInput, first: Int, after: String, last: Int, before: String): AdminEntityConnection
  entityHistory(id: ID!): EntityHistory
  publishHistory(id: ID!): PublishHistory
}

type Mutation {
  publishEntities(entities: [EntityVersionInput!]!): [EntityPublishPayload!]
  unpublishEntities(ids: [ID!]!): [EntityPublishPayload!]
  archiveEntity(id: ID!): EntityPublishPayload
  unarchiveEntity(id: ID!): EntityPublishPayload
  createFooEntity(entity: AdminFooCreateInput!): AdminFoo
  updateFooEntity(entity: AdminFooUpdateInput!): AdminFoo
  createBarEntity(entity: AdminBarCreateInput!): AdminBar
  updateBarEntity(entity: AdminBarUpdateInput!): AdminBar
}
"
`;

exports[`Value type schema spec Generated QL schema 1`] = `
"interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String!
  endCursor: String!
}

enum EntityType {
  Foo
  Bar
}

interface Entity implements Node {
  id: ID!
  _name: String!
}

enum ValueType {
  ValueOne
  ValueList
  NestedValue
}

interface Value {
  _type: ValueType!
}

type RichText {
  blocksJson: String!
  entities: [Entity]
}

type Location {
  lat: Float!
  lng: Float!
}

input LocationInput {
  lat: Float!
  lng: Float!
}

input BoundingBoxInput {
  minLat: Float!
  maxLat: Float!
  minLng: Float!
  maxLng: Float!
}

type Foo implements Node & Entity {
  id: ID!
  _name: String!
  valueOne: ValueOne
  unspecifiedValue: Value
  valueOneOrList: _ValueListOrValueOne
  nestedValue: NestedValue
}

type Bar implements Node & Entity {
  id: ID!
  _name: String!
}

type ValueOne implements Value {
  _type: ValueType!
  one: String
  two: Bar
  three: Location
}

type ValueList implements Value {
  _type: ValueType!
  one: [String!]
  two: [Bar!]
  three: [Location!]
}

type NestedValue implements Value {
  _type: ValueType!
  text: String
  child: NestedValue
}

enum EntityPublishState {
  draft
  published
  modified
  withdrawn
  archived
}

interface AdminEntity {
  id: ID!
  _name: String!
  _type: EntityType!
  _version: Int!
  _publishState: EntityPublishState!
}

type AdminEntityEdge {
  node: AdminEntity
  cursor: String!
}

type AdminEntityConnection {
  pageInfo: PageInfo!
  edges: [AdminEntityEdge]
  totalCount: Int!
}

input AdminQueryInput {
  entityTypes: [EntityType]
  referencing: ID
  boundingBox: BoundingBoxInput
  order: String
  text: String
}

input AdminReferenceInput {
  id: ID!
}

input EntityVersionInput {
  id: ID!
  version: Int!
}

interface AdminValue {
  _type: ValueType!
}

type AdminRichText {
  blocksJson: String!
  entities: [AdminEntity]
}

input AdminRichTextInput {
  blocksJson: String!
}

type EntityVersionInfo {
  version: Int!
  published: Boolean!
  createdBy: ID!
  createdAt: String!
}

type EntityHistory {
  id: ID!
  versions: [EntityVersionInfo]!
}

type PublishEvent {
  version: Int
  publishedBy: ID!
  publishedAt: String!
}

type PublishHistory {
  id: ID!
  events: [PublishEvent]!
}

type EntityPublishPayload {
  id: ID!
}

type AdminFoo implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
  valueOne: AdminValueOne
  unspecifiedValue: AdminValue
  valueOneOrList: _AdminValueListOrValueOne
  nestedValue: AdminNestedValue
}

input AdminFooCreateInput {
  id: ID
  _type: EntityType
  _name: String!
  valueOneJson: String
  valueOne: AdminValueOneInput
  unspecifiedValueJson: String
  valueOneOrListJson: String
  nestedValueJson: String
  nestedValue: AdminNestedValueInput
}

input AdminFooUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
  valueOneJson: String
  valueOne: AdminValueOneInput
  unspecifiedValueJson: String
  valueOneOrListJson: String
  nestedValueJson: String
  nestedValue: AdminNestedValueInput
}

type AdminBar implements AdminEntity {
  id: ID!
  _type: EntityType!
  _name: String!
  _version: Int!
  _publishState: EntityPublishState!
}

input AdminBarCreateInput {
  id: ID
  _type: EntityType
  _name: String!
}

input AdminBarUpdateInput {
  id: ID!
  _type: EntityType
  _name: String
}

type AdminValueOne implements AdminValue {
  _type: ValueType!
  one: String
  two: AdminBar
  three: Location
}

input AdminValueOneInput {
  _type: ValueType!
  one: String
  two: AdminReferenceInput
  three: LocationInput
}

type AdminValueList implements AdminValue {
  _type: ValueType!
  one: [String!]
  two: [AdminBar!]
  three: [Location!]
}

input AdminValueListInput {
  _type: ValueType!
  one: [String!]
  two: [AdminReferenceInput!]
  three: [LocationInput!]
}

type AdminNestedValue implements AdminValue {
  _type: ValueType!
  text: String
  child: AdminNestedValue
}

input AdminNestedValueInput {
  _type: ValueType!
  text: String
  childJson: String
  child: AdminNestedValueInput
}

enum _ValueListOrValueOne {
  ValueList
  ValueOne
}

enum _AdminValueListOrValueOne {
  AdminValueList
  AdminValueOne
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]
  adminEntity(id: ID!, version: Int): AdminEntity
  adminEntities(ids: [ID!]!): [AdminEntity]
  adminSearchEntities(query: AdminQueryInput, first: Int, after: String, last: Int, before: String): AdminEntityConnection
  entityHistory(id: ID!): EntityHistory
  publishHistory(id: ID!): PublishHistory
}

type Mutation {
  publishEntities(entities: [EntityVersionInput!]!): [EntityPublishPayload!]
  unpublishEntities(ids: [ID!]!): [EntityPublishPayload!]
  archiveEntity(id: ID!): EntityPublishPayload
  unarchiveEntity(id: ID!): EntityPublishPayload
  createFooEntity(entity: AdminFooCreateInput!): AdminFoo
  updateFooEntity(entity: AdminFooUpdateInput!): AdminFoo
  createBarEntity(entity: AdminBarCreateInput!): AdminBar
  updateBarEntity(entity: AdminBarUpdateInput!): AdminBar
}
"
`;
